### ISSUE TITLE
compileJava task is UP-TO-DATE when classpath has changed

### ISSUE THREAD

**Issue by hungvietnguyen** [gradle/gradle#6398]
### Your Environment
Gradle 4.9

### Steps to Reproduce
Sample project: [GradleClasspathBug.zip](https://github.com/gradle/gradle/files/2295003/GradleClasspathBug.zip)

This project uses a local repository (the bug doesn't seem to appear with remote repositories):
```
dependencies {
    implementation "com.android.tools.build.jetifier:jetifier-core:1.0.0-alpha07"
}
repositories {
    maven { url 'repo' }
}
```

I build this project with `./gradlew jar --info -Dorg.gradle.caching.debug=true --build-cache` (caching is not relevant here, I use it only to observe the inputs' hashes).

In the first build, I get:

```
> Task :compileJava
Appending inputPropertyHash for 'classpath' to build cache key: 848322ca35d9be4005d2dcb5d99226ef
```

Then I change the dependency from `1.0.0-alpha07` to `1.0.0-alpha08`. In the second build, I get:

```
> Task :compileJava UP-TO-DATE
Appending inputPropertyHash for 'classpath' to build cache key: 848322ca35d9be4005d2dcb5d99226ef
```

Note that the `classpath`'s key remains the same. This is unexpected.

Side note: The above happens without Kotlin, but when I add Kotlin, I notice the following:
```
> Task :compileKotlin
Appending inputPropertyHash for 'classpath' to build cache key: 88d0fda645af489c814ca840ab986249 (this key is different from the first build, which is expected)
Task ':compileKotlin' is not up-to-date because:
  Input property 'classpath' file /.../GradleClasspathBug/repo/com/android/tools/build/jetifier/jetifier-core/1.0.0-alpha08/jetifier-core-1.0.0-alpha08.jar has changed.
```
For some reason, `compileKotlin` works as expected, even when `compileJava` does not.

### Context
Every time I update a dependency, I would need to delete the outputs as incremental build is broken. This also affects other developers on our larger team who might not be familiar with Gradle. Advising them to delete the outputs is merely a workaround, so it would be great if this issue got resolved.

**Comment by wolfs** [comment]
@hungvietnguyen Are you sure that `jetifier-core:1.0.0-alpha07` and `jetifier-core:1.0.0-alpha08` have a different ABI? If they don't then the `compileJava` task is up-to-date (and rightfully so).

**Comment by hungvietnguyen** [comment]
You are right, I checked again and see that they have the same ABI. Previously, I saw `compileKotlin` reported that `classpath` had changed, so I thought this was a bug in the Java plugin, but now I can see that this is in fact a bug in the Kotlin plugin, I've filed https://youtrack.jetbrains.com/issue/KT-26204.

The real issue that we are having is slightly different, I've filed https://github.com/gradle/gradle/issues/6400 to keep the discussion focused. Thanks!
